### DISCUSSION TITLE
Supported RetroArch commands

### DISCUSSION THREAD

**Discussion by gitGalu** [arianrhodsandlot/nostalgist#40] [General]
On the list of available commands (on the RetroArch page linked in the documentation - https://nostalgist.js.org/apis/send-command/#_top ) there is GAME_FOCUS_TOGGLE mentioned. I checked that it was added to this list recently, but the commit that added this functionality dates back to 2016.

Sending this command seems to have no effect in emulation based on Nostalgist, while it's possible to force this mode pressing a keyboard key (input_game_focus_toggle in the configuration).

I'm wondering what determines whether certain commands are supported - and there are other interesting that do not work, like FAST_FORWARD_HOLD. Does Nostalgist need to be updated, or should the core be updated or recompiled?

**Comment by gitGalu** [comment]
I just found out that there is 
     input_auto_game_focus: '1'
 config option that does what I wanted - toggles game focus mode on start, disabling all the hotkeys and allowing full keyboard access on cores such as PUAE.

**Comment by arianrhodsandlot** [comment]
The commands are sent to the RetroArch core directly without any transformation, so if a command does not work as expected while using a latest version of RetroArch core, it's very likely due to some limitations of the browser or Emscripten.

Luckily, as I tried, `GAME_FOCUS_TOGGLE` and `FAST_FORWARD_HOLD` can be used in browsers.

## `GAME_FOCUS_TOGGLE`
It seems to be using the DOM method `requestPointerLock` under the hood, which requires user's interaction. Here is an example:
```js
nostalgist.getCanvas().addEventListener('click', () => {
  nostalgist.sendCommand('GAME_FOCUS_TOGGLE')
});
```

## `FAST_FORWARD_HOLD`
Unlike `FAST_FORWARD`, it needs to be called continuously to take effect. Wrapping it into a `while` loop should work.
```js
while (true) {
  await new Promise((resolve) => setTimeout(resolve, 1))
  nostalgist.sendCommand('FAST_FORWARD_HOLD')
}
```

**Comment by gitGalu** [comment]
Thank you!
